mirgecom/mpi.py

```diff
@@ -32,6 +32,64 @@
 import sys
 
 
+def _check_gpu_oversubscription():
+    """
+    Check whether multiple ranks are running on the same GPU on each node.
+
+    Only works with CUDA devices currently due to the use of the
+    PCI_DOMAIN_ID_NV extension.
+    """
+    from mpi4py import MPI
+    import pyopencl as cl
+
+    size = MPI.COMM_WORLD.Get_size()
+    rank = MPI.COMM_WORLD.Get_rank()
+
+    if size <= 1:
+        return
+
+    cl_ctx = cl.create_some_context()
+    dev = cl_ctx.devices
+
+    # No support for multi-device contexts
+    if len(dev) > 1:
+        raise RuntimeError(f"Multiple devices selected for rank {rank}.")
+
+    dev = dev[0]
+
+    if not (dev.type & cl.device_type.GPU):
+        return
+
+    node_comm = MPI.COMM_WORLD.Split_type(MPI.COMM_TYPE_SHARED)
+    node_rank = node_comm.Get_rank()
+
+    try:
+        domain_id = hex(dev.pci_domain_id_nv)
+    except (cl._cl.LogicError, AttributeError):
+        from warnings import warn
+        warn("Cannot detect whether multiple ranks are running on the"
+             " same GPU because it requires Nvidia GPUs running with"
+             " pyopencl>2021.1.1 and (Nvidia CL or pocl>1.6).")
+    else:
+        bus_id = hex(dev.pci_bus_id_nv)
+        slot_id = hex(dev.pci_slot_id_nv)
+
+        dev_id = (domain_id, bus_id, slot_id)
+
+        dev_ids = node_comm.gather(dev_id, root=0)
+
+        if node_rank == 0:
+            if len(dev_ids) != len(set(dev_ids)):
+                hostname = MPI.Get_processor_name()
+                dup = [item for item in dev_ids if dev_ids.count(item) > 1]
+
+                raise RuntimeError(
+                      f"Multiple ranks are sharing GPUs on node '{hostname}'."
+                      f" Duplicate PCIe IDs: {dup}.")
+
+    node_comm.Free()
```

Comment:
This alloc/free pair should be a context manager.

Reply:
I didn't find much info regarding mpi4py context managers so I implemented a simple one.